app/containers/LoginCallback/sagas.js

```diff
@@ -0,0 +1,27 @@
+/**
+ * 1. sets up auth library and waits for auth token
+ * 2. gets user profile
+ */
+
+import { call, put } from 'redux-saga/effects';
+import { browserHistory } from 'react-router';
+import { loginCallback, getProfile } from './lib';
+import { loginSuccess, loginError } from './actions';
+
+function* loginCallbackSaga() {
+  // TODO take the proper event so that this saga is run exactly once (when loading route)
```

Comment:
@sourcesoft thanks! Didn't think of `cancel`. I tried just doing
```
    const task = yield call(loginCallbackSaga);
    yield cancel(task);
```

But this whole code will still be executed again, which defeats the purpose. I couldn't use `fork(takeLatest, LOCATION_CHANGE, loginCallbackSaga)` either, it seems it doesn't receive the `LOCATION_CHANGE` action. And anyway I see in the logs that `LOCATION_CHANGE` is fired twice when doing the auth callback, one initial time coming from a `null` prev state to `/login/callback` next state, but also immediately a second time going from `/login/callback` to `/login/callback`. The only difference is that in the prev state, the `hash` is populated (with the auth idToken and state query params), and in the next state it's empty. I think auth0 must be doing a redirect onto itself as it's parsing this data, then when done it passes this parsed data to the `lock.on('authenticated', ...)` callback.

So I think I need to create and listen to an event that would be fired once `lock.on('authenticated', ...)` callback is called. (another way would be to somehow make the LOCATION_CHANGE listener work, and check whether the hash is empty, but that would be really hacky)